nav2_util/include/nav2_util/twist_subscriber.hpp

```diff
@@ -0,0 +1,92 @@
+// Copyright (C) 2023 Ryan Friedman
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+
+#ifndef NAV2_UTIL__TWIST_SUBSCRIBER_HPP_
+#define NAV2_UTIL__TWIST_SUBSCRIBER_HPP_
+
+
+#include <memory>
+#include <string>
+#include <utility>
+
+#include "geometry_msgs/msg/twist.hpp"
+#include "geometry_msgs/msg/twist_stamped.hpp"
+#include "rclcpp/parameter_service.hpp"
+#include "rclcpp/rclcpp.hpp"
+#include "rclcpp/qos.hpp"
+#include "rclcpp_lifecycle/lifecycle_publisher.hpp"
+
+#include "lifecycle_node.hpp"
+#include "node_utils.hpp"
+
+namespace nav2_util
+{
+
+/**
+ * @class nav2_util::TwistSubscriber
+ * @brief A simple wrapper on a Twist subscriber that receives either Twist or TwistStamped
+ *
+ * The default is to subscribe to Twist to preserve backwards compatibility, but it can be overridden
+ * using the "enable_stamped_cmd_vel" parameter to subscribe to TwistStamped.
+ *
+ */
+
+class TwistSubscriber
+{
+public:
+  /**
+  * @brief A constructor
+  * @param nh The node
+  * @param topic publisher topic name
+  * @param qos publisher quality of service
+  */
+  template<typename TwistCallbackT,
+    typename TwistStampedCallbackT
+  >
+  explicit TwistSubscriber(
+    nav2_util::LifecycleNode::SharedPtr node,
+    const std::string & topic,
+    const rclcpp::QoS & qos,
+    TwistCallbackT && TwistCallback,
+    TwistStampedCallbackT && TwistStampedCallback
+  )
+  {
+    using nav2_util::declare_parameter_if_not_declared;
+    declare_parameter_if_not_declared(
+      node, "enable_stamped_cmd_vel",
+      rclcpp::ParameterValue{false});
+    node->get_parameter("enable_stamped_cmd_vel", is_stamped_);
+    if (is_stamped_) {
+      twist_stamped_sub_ = node->create_subscription<geometry_msgs::msg::TwistStamped>(
+        topic,
+        qos,
+        std::forward<TwistStampedCallbackT>(TwistStampedCallback));
+    } else {
+      twist_sub_ = node->create_subscription<geometry_msgs::msg::Twist>(
+        topic,
+        qos,
+        std::forward<TwistCallbackT>(TwistCallback));
+    }
+  }
```

Comment:
Q: why not have a callback here and a `getTwist` & `getTwistStamped` member?

Reply:
I didn't know what the behavior of getTwistStamped would be if the node received Twist. Does it re-stamp the timestamp automagically? Right now, there is no made up timestamps. Either it has timestamp data, or it doesn't and you aren't allowed to use it.

Reply:
Could throw?

I'm not necessarily asking for a change, just a question to see if it would be sensible to make this similar the `OdomSmoother` or other subscriber-data objects which hold their data and get when requested. I suppose often the Twist Subscription clients are those that need event-based access to the twist data, not polling based, which would be a reasonable answer for why this method is necessary as well.

Reply:
> Could throw?
> 
> I'm not necessarily asking for a change, just a question to see if it would be sensible to make this similar the `OdomSmoother` or other subscriber-data objects which hold their data and get when requested. I suppose often the Twist Subscription clients are those that need event-based access to the twist data, not polling based, which would be a reasonable answer for why this method is necessary as well. So perhaps its just not sensible?

I think it could be sensible, but I worry that it is a bit potential higher impact (probability of bugs). Without sufficient test infrastructure to run the whole system tests with a general parameter override globally for the new param, I'm hesitant to introduce data-sharing modifications.